mungegithub/mungers/publish_scripts/util.sh

```diff
@@ -0,0 +1,119 @@
+#!/bin/bash
+
+# Copyright 2017 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This file includes functions shared by the each repository's publish scripts.
+
+# sync_repo() cherry picks the latest changes in k8s.io/kubernetes/${filter} to the
+# local copy of the repository to be published.
+#
+# prerequisites
+# 1. we are in the root of the repository to be published
+# 2. we are on the branch to be published (let's call it "target-branch")
+# overall flow
+# 1. fetch the current level of k8s.io/kubernetes
+# 2. check out the $src_branch of k8s.io/kubernetes as branch kube-sync
+# 3. rewrite the history of branch kube-sync to *only* include code in $subdirectory
+# 4. locate all commits between the last time we sync'ed and now
+# 5. switch back to the "target-branch"
+# 6. for each commit, cherry-pick it (which will keep authorship) into "target-branch"
+# 7. update metadata files indicating which commits we've sync'ed to
+sync_repo() {
+    # subdirectory in k8s.io/kubernetes, e.g., staging/src/k8s.io/apimachinery
+    local subdirectory="${1}"
+    local src_branch="${2}"
+    readonly filter src_branch
+
+    local currBranch=$(git rev-parse --abbrev-ref HEAD)
+    local previousKubeSHA=$(cat kubernetes-sha)
+    local previousBranchSHA=$(cat filter-branch-sha)
+    
+    git remote add upstream-kube https://github.com/kubernetes/kubernetes.git || true
+    git fetch upstream-kube
+    git branch -D kube-sync || true
+    git checkout upstream-kube/"${src_branch}" -b kube-sync
+    git reset --hard upstream-kube/"${src_branch}"
+    local newKubeSHA=$(git log --oneline --format='%H' kube-sync -1)
+    
+    # this command rewrites git history to *only* include $subdirectory 
+    git filter-branch -f --subdirectory-filter "${subdirectory}" HEAD
```

Comment:
@caesarxuchao For our OpenShift downstream work we have to be able to cherry-pick commits over multiple repos, i.e. apimachinery, client-go, kube. It would help **a lot** to have the upstream commit hash in the commit msg. This can easily be done like this:

```
git filter-branch -f --msg-filter 'awk 1 - && echo && echo "Kubernetes-commit: ${GIT_COMMIT}"' --subdirectory-filter "${subdirectory}" HEAD
```

Note: `awk 1 -` sanitizes the trailing newline, compare http://backreference.org/2010/05/23/sanitizing-files-with-no-trailing-newline/